### PR TITLE
fix(VSlideGroup): detect horizontal/vertical swipe intent

### LINKED ISSUE #10673: [Feature Request] v-slide-group momentum and scroll lock
### Problem to solve
On mobile touch based devices, if you try swiping left and right the experience is not ideal.  First, there's no momentum when you let go -- the illusion of physical weight.  Second, as you try to swipe through content horizontally, finger movements in the vertical direction cause the screen to scroll up and down unexpectedly.

### Proposed solution
Once horizontal swipe intent is determined, vertical scroll passthrough should be temporarily disabled.  Add realistic gestures like momentum scrolling:  https://material.io/design/interaction/gestures.html

### LINKED ISSUE #12293: [Bug Report] Mobile Touch: cannot swipe vertically when in contact with vuetify slider group
### Environment
**Vuetify Version:** 2.3.10
**Vue Version:** 2.6.12
**Browsers:** Chrome 87.0.4270.0
**OS:** Windows 10

### Steps to reproduce
1. Create slider group (Codepen link added) 
2. Create additional content (long enough to scroll the page)
2. Use touch supported device or mobile view in browser
3. Scroll/swipe through y-axis of the slide group (vertically)

### Expected Behavior
The page must be able to scroll up/down

### Actual Behavior
The swipe is inherited by the slide group, making it prevent page scrolling.

### Reproduction Link
<a href="https://codepen.io/sdriyaz712/pen/abNMrMR" target="_blank">https://codepen.io/sdriyaz712/pen/abNMrMR</a>

### PR DESCRIPTION
**Pull request by yuwu9145**
fixes #12293

partically fixes #10673

## Description

As per suggestions in #10673, `VSlideGroup` should determine swipe intent (horizontal or vertical) first, then using swipe intent to only allow one of horizontal/vertical swipe gestures between each `onTouchStart` and `onTouchEnd` session

fixes #12293

partically fixes #10673

## Motivation and Context

fixes #12293

partically fixes #10673

## How Has This Been Tested?

visually

## Markup:

<details>

```vue
// Paste your FULL Playground.vue here
<template>
  <v-container>
    <v-slide-group>
      <v-slide-item v-for="i in 10" :key="i" class="ma-5">
        <v-card color="#eeeeee" elevation="10" max-width="300px">
          <v-img
            src="https://cdn.vuetifyjs.com/images/cards/sunshine.jpg"
            height="200px"
          ></v-img>
          <v-card-text>
            SWITCH TO MOBILE
            <v-divider></v-divider>
            Try swiping up/down while still keeping finger on these slides
          </v-card-text>
          <v-card-actions>
            <v-spacer />
            <v-btn text> Go </v-btn>
          </v-card-actions>
        </v-card>
      </v-slide-item>
    </v-slide-group>
    <v-card class="ma-2 pa-5" v-for="i in 40" :key="i">
      <v-content>You cannot scroll past the above slides on mobile</v-content>
    </v-card>
  </v-container>
</template>

<script>
export default {
  data() {
    return {
    };
  },
};
</script>
```
</details>

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and backwards compatible changes and `next` for non-backwards compatible changes).
- [x] My code follows the code style of this project.
- [ ] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)

### CODE REVIEW

File: packages/vuetify/src/components/VSlideGroup/VSlideGroup.ts [hunk 1]
```diff
@@ -287,7 +291,20 @@ export const BaseSlideGroup = mixins<options &
...
+        const diffX = e.touchmoveX - e.touchstartX
+        const diffY = e.touchmoveY - e.touchstartY
+        this.isSwipingHorizontal = Math.abs(diffX) > Math.abs(diffY)
+        this.isSwiping = true
+      }
+      if (this.isSwipingHorizontal) {
+        // sliding horizontally
+        this.scrollOffset = this.startX - e.touchmoveX
+        // temporarily disable window vertical scrolling
+        document.documentElement.style.overflowY = 'hidden'
```
Comment (yuwu9145) [inline packages/vuetify/src/components/VSlideGroup/VSlideGroup.ts:306]: When swiping horizontally, I tried to achieve disabling vertical scrolling or touch swipe via css by setting `touch-action` to `pan-x`. However, `touch-action` only takes effect from next touch swipe and changing `touch-action` to `pan-x` will not restrict current touch move to "horizontal only". So I use javascript way that disables whole page's vertical scroll if swiping horizontally. (And remove this style in `touchEnd`). Please advice if there is a better way to disable vertical scroll passthrough once horizontal swipe intent is detected

**Comment by johnleider** [comment]
@babyraging can you reproduce your issue in #11571 with this?

**Comment by yuwu9145** [comment]
> @babyraging can you reproduce your issue in #11571 with this?

Hi @johnleider , I did a test simulating @babyraging issue in #11571, this pull request does resolve it.



This is the mock that I used to test:
<details>

```vue
<template>
  <v-container>
      <v-chip-group
          mandatory
          column
          active-class="primary--text"
        >
          <v-chip
            v-for="tag in tags"
            :key="tag"
          >
            {{ tag }}
          </v-chip>
        </v-chip-group>

  </v-container>
</template>

<script>
export default {
  data() {
    return {
    tags: [
        'Work',
        'Home Improvement',
        'Vacation',
        'Food',
        'Drawers',
        'Shopping',
        'Art',
        'Tech',
        'Creative Writing',
        'Shopping',
        'Art',
        'Tech',
        'Shopping',
        'Art',
        'Tech',
        'Shopping',
        'Art',
        'Tech',
        'Shopping',
        'Art',
        'Tech',
        'Work',
        'Home Improvement',
        'Vacation',
        'Food',
        'Drawers',
        'Shopping',
        'Art',
        'Tech',
        'Creative Writing',
        'Shopping',
        'Art',
        'Tech',
        'Shopping',
        'Art',
        'Tech',
        'Shopping',
        'Art',
        'Tech',
        'Shopping',
        'Art',
        'Tech',
        'Work',
        'Home Improvement',
        'Vacation',
        'Food',
        'Drawers',
        'Shopping',
        'Art',
        'Tech',
        'Creative Writing',
        'Shopping',
        'Art',
        'Tech',
        'Shopping',
        'Art',
        'Tech',
        'Shopping',
        'Art',
        'Tech',
        'Shopping',
        'Art',
        'Tech',
        'Work',
        'Home Improvement',
        'Vacation',
        'Food',
        'Drawers',
        'Shopping',
        'Art',
        'Tech',
        'Creative Writing',
        'Shopping',
        'Art',
        'Tech',
        'Shopping',
        'Art',
        'Tech',
        'Shopping',
        'Art',
        'Tech',
        'Shopping',
        'Art',
        'Tech',
        'Work',
        'Home Improvement',
        'Vacation',
        'Food',
        'Drawers',
        'Shopping',
        'Art',
        'Tech',
        'Creative Writing',
        'Shopping',
        'Art',
        'Tech',
        'Shopping',
        'Art',
        'Tech',
        'Shopping',
        'Art',
        'Tech',
        'Shopping',
        'Art',
        'Tech',
        'Work',
        'Home Improvement',
        'Vacation',
        'Food',
        'Drawers',
        'Shopping',
        'Art',
        'Tech',
        'Creative Writing',
        'Shopping',
        'Art',
        'Tech',
        'Shopping',
        'Art',
        'Tech',
        'Shopping',
        'Art',
        'Tech',
        'Shopping',
        'Art',
        'Tech',
      ],
    };
  }
};
</script>

```
</details>

Also, @babyraging please feel free to leave any comments

**Comment by HammerMeetNail** [comment]
Eagerly awaiting this fix 🙏 

Recently deployed a few slides on a single page and it’s very difficult to scroll the page on mobile. If you need someone to try this out to confirm it fixes vertical scroll on mobile I should be able to verify on my site tomorrow night.

File: packages/vuetify/src/components/VSlideGroup/VSlideGroup.ts [hunk 2]
```diff
@@ -287,7 +291,20 @@ export const BaseSlideGroup = mixins<options &
       content.style.setProperty('willChange', 'transform')
     },
     onTouchMove (e: TouchEvent) {
```
[the remark this answers is not in the document]
Comment (johnleider) [reply packages/vuetify/src/components/VSlideGroup/VSlideGroup.ts:293]: In addition, you will need to import IN_BROWSER from '@/globals'

File: packages/vuetify/src/components/VSlideGroup/VSlideGroup.ts [hunk 3]
```diff
@@ -311,6 +328,10 @@ export const BaseSlideGroup = mixins<options &
           this.scrollOffset = maxScrollOffset
         }
       }
+
+      this.isSwiping = false
+      // rollback whole page scrolling to default
+      document.documentElement.style.removeProperty('overflow-y')
```
Comment (johnleider) [inline packages/vuetify/src/components/VSlideGroup/VSlideGroup.ts:334]: Same here, need to check if IN_BROWSER

**Review by johnleider — CHANGES_REQUESTED** [review]

File: packages/vuetify/src/components/VSlideGroup/VSlideGroup.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (yuwu9145) [reply packages/vuetify/src/components/VSlideGroup/VSlideGroup.ts:301]: Hi @johnleider , 

looks like `import { IN_BROWSER } from '@/globals'` is only available in `packages/docs`. In `packages/vuetify`, components are doing individual `typeof window !== 'undefined'`, for example in line 46 of `packages/vuetify/src/mixins/scrollable/index.ts`. Therefore, I made a similar `canTouch` computed property to make sure touch event functions `onTouchMove` and `onTouchEnd` are only triggered in web-browser environment.

File: packages/vuetify/src/components/VSlideGroup/VSlideGroup.ts (hunk 3 above)
Comment (yuwu9145) [reply packages/vuetify/src/components/VSlideGroup/VSlideGroup.ts:347]: I guess the whole `onTouchEnd` function should be guarded and checked by `typeof window !== 'undefined'` ? So `if (!this.canTouch) return ` is added to the begginning of this function in line 321

File: packages/vuetify/src/components/VSlideGroup/VSlideGroup.ts [hunk 4]
```diff
@@ -101,6 +101,14 @@ export const BaseSlideGroup = mixins<options &
   }),
 
   computed: {
+    /**
+     * A computed property that returns
+     * whether touch features is
+     * enabled or disabled
+     */
```
Comment (johnleider) [inline packages/vuetify/src/components/VSlideGroup/VSlideGroup.ts:108]:
```suggestion
```

Comment (johnleider) [reply packages/vuetify/src/components/VSlideGroup/VSlideGroup.ts:108]: Not necessary

**Review by johnleider — CHANGES_REQUESTED** [review]
I'll take a look at the globals issue. You won't be able to import using the `@` syntax.

**Comment by yuwu9145** [comment]
> I'll take a look at the globals issue. You won't be able to import using the `@` syntax.

Hi @johnleider , just want to clarify, should I add `import { IN_BROWSER } from '../../../../docs/src/util/globals'` into `VSliderGroup`? Feeling it is pretty hecky though (no component currently does this) and it results in compile error

**Comment by HammerMeetNail** [comment]
Hi all - super appreciative of the work you're all doing. Looking to give this a bump cause it should make our site a lot more usable on mobile.

**Comment by yuwu9145** [comment]
> Hi all - super appreciative of the work you're all doing. Looking to give this a bump cause it should make our site a lot more usable on mobile.

Would you mind helping to verify if this fix resolves your website issue? I am happy to give you a guidance of how to do it if you need help ;)

**Comment by HammerMeetNail** [comment]
Absolutely, I'll have time tomorrow night EDT. I'll give you a shout if I have any trouble.

**Comment by HammerMeetNail** [comment]
@yuwu9145 I can confirm this fixes our slide group issue. Our page looks roughly like this: 

```
<template>
  <div class="d-flex flex-grow-1 flex-column">

    <!-- First Slide Group-->
    <header>Header1</header>
    <v-row class="flex-grow-0" dense>
      <v-col cols="12">
        <slide :stuff="stuff"></slide>
      </v-col>
    </v-row>
    
    <!-- Second Slide Group-->
    <header>Header2</header>
    <v-row class="flex-grow-0" dense>
      <v-col cols="12">
        <slide :stuff="stuff"></slide>
      </v-col>
    </v-row>
    
    <!-- Third Slide Group-->
    <header>Header3</header>
    <v-row class="flex-grow-0" dense>
      <v-col cols="12">
        <slide :stuff="stuff"></slide>
      </v-col>
    </v-row>

  </div>
</template>
```

Before your fix we'd be able to recreate the issue in Chrome dev tools with the following steps:

- Open inspector
- Switch to mobile view
- Try and click an item in the slide group and drag (just like you'd do if you were using touch)

This normally would prevent vertical scroll. With your fix, we can click any item in the slide group and scroll vertically. 

Very eager to see this PR touched up and merged :)

**Comment by yuwu9145** [comment]
@HammerMeetNail Thanks a lot!

**Comment by yuwu9145** [comment]
@johnleider any plan to further review this ?

**Review by johnleider — APPROVED** [review]

**Review by johnleider — APPROVED** [review]

**Comment by FrostySnot** [comment]
Gage waz here

**Comment by FrostySnot** [comment]
> Gage waz here

**Comment by FrostySnot** [comment]
Tewa

**Comment by mahammad-sixberries** [comment]
temporary fix 
```
.v-slide-group__wrapper {
    touch-action: auto;
    overflow: scroll;
    scroll-behavior: smooth;
    scroll-snap-type: x mandatory;
}
```

**Pull request merged**